editoast/src/views/timetable/stdcm/failure_handler.rs

```diff
@@ -0,0 +1,125 @@
+use std::sync::Arc;
+
+use chrono::DateTime;
+use chrono::Utc;
+
+use crate::core::conflict_detection::ConflictDetectionRequest;
+use crate::core::conflict_detection::WorkSchedulesRequest;
+use crate::core::simulation::SimulationResponse;
+use crate::core::AsCoreRequest;
+use crate::core::CoreClient;
+use crate::error::Result;
+use crate::models::train_schedule::TrainSchedule;
+use crate::models::work_schedules::WorkSchedule;
+use crate::views::timetable::stdcm::StdcmResponse;
+
+use super::build_train_requirements;
+use super::VirtualTrainRun;
+
+/// `SimulationFailureHandler` is used when a simulation failure occurs,
+/// particularly when a train's path cannot be found. It helps detect
+/// conflicts between the `virtual_train` and existing train schedules and simulations.
+/// `virtual_train` is a simulated train created to detect conflicts
+/// when a real train’s path cannot be found during the simulation.
+pub struct SimulationFailureHandler {
+    pub(super) core_client: Arc<CoreClient>,
+    pub(super) infra_id: i64,
+    pub(super) infra_version: String,
+    pub(super) train_schedules: Vec<TrainSchedule>,
+    pub(super) simulations: Vec<SimulationResponse>,
+    pub(super) work_schedules: Vec<WorkSchedule>,
+    pub(super) virtual_train_run: VirtualTrainRun,
+    pub(super) earliest_departure_time: DateTime<Utc>,
+    pub(super) latest_simulation_end: DateTime<Utc>,
+}
+
+impl SimulationFailureHandler {
+    // Filter the provided work schedules to find those that conflict with the given parameters
+    // This identifies any work schedules that may overlap with the earliest departure time and latest simulation end.
+    fn get_work_schedules_request(&self) -> Option<WorkSchedulesRequest> {
+        if self.work_schedules.is_empty() {
+            return None;
+        }
+        let work_schedule_requirements = self
+            .work_schedules
+            .iter()
+            .filter_map(|ws| {
+                if let Some(core_ws) = ws
+                    .as_core_work_schedule(self.earliest_departure_time, self.latest_simulation_end)
+                {
+                    return Some((ws.id, core_ws));
+                };
+                None
+            })
+            .collect();
+
+        Some(WorkSchedulesRequest {
+            start_time: self.earliest_departure_time,
+            work_schedule_requirements,
+        })
+    }
+
+    pub(super) async fn compute_conflicts(self) -> Result<StdcmResponse> {
+        let work_schedules = self.get_work_schedules_request();
+        let Self {
+            mut train_schedules,
+            mut simulations,
+            virtual_train_run,
+            infra_id,
+            infra_version,
+            earliest_departure_time,
+            latest_simulation_end,
+            ..
+        } = self;
+        let VirtualTrainRun {
+            train_schedule,
+            simulation,
+            pathfinding,
+        } = virtual_train_run;
```

Comment:
```suggestion
        let Self {
            mut train_schedules,
            mut simulations,
            virtual_train_run:
                VirtualTrainRun {
                    train_schedule,
                    simulation,
                    pathfinding,
                },
            infra_id,
            infra_version,
            earliest_departure_time,
            latest_simulation_end,
            ..
        } = self;

```

You can destructure several levels at once.

Reply:
Done.